### DISCUSSION TITLE
Easier/more documented way to use jpegli on Linux

### DISCUSSION THREAD

**Discussion by Commenter25** [libvips/libvips#4177] [Q&A] [answered]
> "That's a bit cumbersome, so if you think a better way would be nice, maybe worth opening an issue?"
> 	- https://github.com/libvips/libvips/issues/3872#issuecomment-2040265103

I've spent the past week or so (off and on) trying to figure out how to compile vips in a way that would use Jpegli. For the first time in probably 4 years, I have wished I was developing on Windows, so I could just use the [build container](https://github.com/libvips/build-win64-mxe) and move on with my life. I don't know if this is even anyone's fault other than mine; in fact, I'm almost certain it's entirely my fault, and that I'm missing something incredibly obvious; but I feel like I have to say something somewhere, because I need to sleep, and I feel like I need to find a way to stop thinking about this. I may realize what I've been doing wrong in the morning and just be completely embarrassed by posting this. I don't know.

I frankly have basically no experience compiling and installing things from compiled source because I just avoid it as much as I can because the experience drives me mad, but I'm insistent on having optimized images, so now I'm here. Attempting to wrap my brain around compilation and shared libraries is not the adventure I expected to go on in my optimization hyperfixation. And yet I went down that path, and have found no fruits. With how little I have found about compiling with Jpegli, it feels like I'm the only person in the world who has done this before, even though I know that is preposterous. Either that, or it's extremely simple, to the point nobody has felt the need to spell it out. I have no idea what I'm missing, and I desperately wish to forget all of this and move on to more fun things, like groaning at CSS instead of my terminal.

It is the midnight of a new month. This is a broken mind waving the white flag. Please save me.

**Comment by jcupitt** [comment]
Hello @Commenter25,

I'm sorry you've lost so much hair over this! 

I've not tried this myself, but my understanding is that jpegli is a drop-in binary replacement for the usual libjpeg library. So I would try building it to some prefix, then setting `LD_LIBRARY_PATH` so that the `lib/libjpeg.so` from jpegli overrides the usual `libjpeg.so`.

Something like mozjpeg, which is not API-compatible, is a LOT harder to replace. But jpegli ought to be pretty simple (hopefully).

**Comment by jcupitt** [comment, marked as answer]
I had a quick go. 

If I check my existing binaries, they are linked to the API 8.2.2 libjpeg (this is Ubuntu 24.04):

```
$ ldd $(which vips) | grep jpeg
	libjpeg.so.8 => /lib/x86_64-linux-gnu/libjpeg.so.8 (0x00007014a8c8f000)
$ ls /lib/x86_64-linux-gnu/libjpeg.so*
/lib/x86_64-linux-gnu/libjpeg.so    /lib/x86_64-linux-gnu/libjpeg.so.8.2.2
/lib/x86_64-linux-gnu/libjpeg.so.8
```

In libjxl I checked out v0.11 and set `JPEGLI_LIBJPEG_LIBRARY_VERSION` to 8.2.2 and `JPEGLI_LIBJPEG_LIBRARY_SOVERSION` to 8, then installed.

It doesn't install the jpegli binary in the install prefix, it leaves it in the libjxl build area, I suppose to avoid too much confusion:

```
john@banana ~/GIT/libjxl/build ((HEAD detached at v0.11.0)) $ find . -name "*libjpeg*"
./lib/jpegli/libjpeg.so
./lib/jpegli/libjpeg.so.8
./lib/jpegli/libjpeg.so.8.2.2
./lib/CMakeFiles/jpegli-libjpeg-obj.dir
./lib/CMakeFiles/jpegli-libjpeg-obj.dir/jpegli/libjpeg_wrapper.cc.o
./lib/CMakeFiles/jpegli-libjpeg-obj.dir/jpegli/libjpeg_wrapper.cc.o.d
./lib/libjpegli-static.a
./tools/libjpegli_jni.so
```

So I can run my existing binaries against the jpegli library like this:

```
$ LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/john/GIT/libjxl/build/lib/jpegli ldd $(which vips) | grep jpeg
	libjpeg.so.8 => /home/john/GIT/libjxl/build/lib/jpegli/libjpeg.so.8 (0x00007da12a8f1000)
```

I tried a benchmark:

```
$ LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/john/GIT/libjxl/build/lib/jpegli /usr/bin/time vips copy wtc.jpg x-jpegli.jpg
1.81user 0.22system 0:01.20elapsed 168%CPU (0avgtext+0avgdata 212736maxresident)k
0inputs+24872outputs (0major+44708minor)pagefaults 0swaps

$ /usr/bin/time vips copy wtc.jpg x-jpeg-turbo.jpg
1.01user 0.18system 0:00.69elapsed 173%CPU (0avgtext+0avgdata 204288maxresident)k
0inputs+29592outputs (0major+42381minor)pagefaults 0swaps

$ ls -l x* 
-rw-r--r-- 1 john john   12733336 Oct  1 12:48 x-jpegli.jpg
-rw-r--r-- 1 john john   15149568 Oct  1 12:49 x-jpeg-turbo.jpg
```

So jpegli is about half the speed of libjpeg-turbo, but the files are 15% smaller.

**Comment by jcupitt** [comment]
Some more background:

## Recompiling

It's really hard to get libvips to use mozjpeg. This version of libjpeg is *not* binary compatible, so you have to recompile to use it. Sadly, almost all packages depend on libjpeg, directly or indirectly,  so you end up having to rebuild almost the whole world. 

I made an example dockerfile that does this a while ago:

https://github.com/jcupitt/docker-builds/blob/master/libvips-mozjpeg-ubuntu23.10/Dockerfile

You have to turn off a lot of libvips dependencies, and it'd be a lot of work to reenable them.

## Drop-in ABI replacements

The pain is so awful that most libjpeg replacements aim for *ABI compatibility*, they carefully and exactly match the ABI (the binary interface) of the old libjpeg. 

Now all you need to do is copy the new library into your lib area, and everything will automatically use the new library.

```
$ sudo cp my-new-libjpeg.so /lib/x86_64-linux-gnu/libjpeg.so.8
```

That would work, but no sane linux person wants to write to `/lib/x86_64-linux-gnu`. That directory is managed by the system and you *must not touch it* or you'll probably break your PC.

Fortunately, the linux linker has a feature where you can override the link search path. `LD_LIBRARY_PATH` lists a set of directories that should be searched for libraries before the standard ones. You can set this to swap libraries at runtime.

**Comment by kleisauke** [comment]
> It doesn't install the jpegli binary in the install prefix, it leaves it in the libjxl build area

Building with `-DJPEGXL_INSTALL_JPEGLI_LIBJPEG=ON` ought to resolve this, see for example:
https://github.com/libvips/build-win64-mxe/blob/3235b79110cf4ddbba54a2a8e1c234ac64e70a0c/build/plugins/jpegli/jpegli.mk#L25

**Comment by Commenter25** [comment]
I must apologize for leaving my late-night rambles of pain into your issue tracker. That said, I'm glad I did, because your explanations did indeed solve my problems! And I'm sure this thread's existence will help someone else too ^^

I was under the impression I would need to completely recompile vips, ImageMagick, and whatever. Likely due to seeing discussions around mozjpeg, and the details you described. I was attempting to do this referencing that dockerfile you mentioned, the Windows build system, skimming a lot of build options and such; only for me to run into constant weird conflicts and errors...

...but it does turn out I was *massively* overcomplicating it! Simply compiling JXL and throwing the libraries into `~/.local/lib` (and binaries into `~/.local/bin`) worked without any hassle. Note that, at least for me on Fedora, I did need to run `sudo ldconfig` for it to recognize them. Additionally, my vips binaries (from Fedora repos) were linked to libjpeg62, so I didn't need the options to build for libjpeg8.

I will mention, mozjpeg also describes itself as ABI compatible:
> MozJPEG is compatible with the libjpeg API and ABI. It is intended to be a drop-in replacement for libjpeg. MozJPEG is a strict superset of libjpeg-turbo's functionality. All MozJPEG's improvements can be disabled at run time, and in that case it behaves exactly like libjpeg-turbo.

and later says "There's no need to write any MozJPEG-specific integration code" in the features list. Because of that, I wouldn't have guessed there was any distinction. I'm still not sure I entirely understand. My best guess is that "superset of functionality" part, in that it's *technically* true when everything's disabled, but does require recompiling for anything that makes mozjpeg special. It's not a big deal to me, as I had no plans to use mozjpeg.

**Reply by jcupitt** [reply]
I'm probably a bit out of date. I think mozjpeg and libjpeg-turbo now have an ABI-compatible system for managing extensions, but they are not compatible with the original IJG libjpeg. Anyway, I'm glad you fixed it.
